core/src/main/scala/kafka/log/Log.scala

```diff
@@ -565,8 +565,8 @@ class Log(val dir: File,
   def deleteOldSegments(predicate: LogSegment => Boolean): Int = {
     // find any segments that match the user-supplied predicate UNLESS it is the final segment
     // and it is empty (since we would just end up re-creating it
-    val lastSegment = activeSegment
-    val deletable = logSegments.takeWhile(s => predicate(s) && (s.baseOffset != lastSegment.baseOffset || s.size > 0))
+    val lastEntry = segments.lastEntry
+    val deletable = logSegments.takeWhile(s => predicate(s) && (lastEntry != null && s.baseOffset != lastEntry.getValue.baseOffset || s.size > 0))
```

Comment:
The reason for synchronizing properly before LogCleaner and Log.delete() is that LogCleaner can create new files after a log dir is removed, which would be bad. 

Thinking about this a bit more, deleteOldSegments() can also add new files due to the call to roll(). So, it would be good to synchronize this properly with Log.delete() as @ijuma suggested. Since deleteOldSegments() is only called infrequently from the scheduler, it seems that we could just take the original approach, i.e., synchronize the whole deleteOldSegments(). There shouldn't be much overhead added due to lock contention.